DataStructures/MinMaxHeap.cs

```diff
@@ -0,0 +1,399 @@
+                }
+            }
+
+            return resIndex;
+        }
+
+        private int GetMaxNodeIndex()
+        {
+            return Count switch
+            {
+                0 => throw new InvalidOperationException("Heap is empty"),
+                1 => 0,
+                2 => 1,
+                _ => Comparer.Compare(Heap[1], Heap[2]) > 0 ? 1 : 2
+            };
+        }
+
+        private bool HasChild(int index) => (index * 2 + 1) < Count;
+
+        private bool IsGrandchild(int node, int grandchild) => grandchild > 2 && Grandparent(grandchild) == node;
+
+        /// <summary>
+        /// Checks if node at index belongs to Min or Max level of the heap.
+        /// Root node belongs to Min level, its children - Max level,
+        /// its grandchildren - Min level, and so on.
+        /// </summary>
+        /// <param name="index">Index to check.</param>
+        /// <returns>true if index is at Min level; false if it is at Max Level.</returns>
+        private bool IsMinLevelIndex(int index)
+        {
+            // For all Min levels, value (index + 1) has the leftmost bit set to '1' at even position.
+            const uint minLevelsBits = 0x55555555;
+            const uint maxLevelsBits = 0xAAAAAAAA;
+            return ((index + 1) & minLevelsBits) > ((index + 1) & maxLevelsBits);
+        }
+
+        private int Parent(int index) => (index - 1) / 2;
+
+        private int Grandparent(int index) => ((index - 1) / 2 - 1) / 2;
+
+        /// <summary>
+        /// Assuming that children sub-trees are valid heaps, pushes node to lower levels
+        /// to make heap valid.
+        /// </summary>
+        /// <param name="index">Node index.</param>
+        private void PushDown(int index)
+        {
+            if (IsMinLevelIndex(index))
+            {
+                PushDownMin(index);
+            }
+            else
+            {
+                PushDownMax(index);
+            }
+        }
+
+        private void PushDownMax(int index)
+        {
+            if (!HasChild(index))
+            {
+                return;
+            }
+
+            int maxIndex = ExtremumChildOrGrandchild(index, minimum: false);
+
+            // If smaller element are put at min level (as result of swaping), it doesn't affect sub-tree validity.
+            // If smaller element are put at max level, PushDownMax() should be called for that node.
+            if (IsGrandchild(index, maxIndex))
+            {
+                if (Comparer.Compare(Heap[maxIndex], Heap[index]) > 0)
+                {
+                    SwapNodes(maxIndex, index);
+                    if (Comparer.Compare(Heap[maxIndex], Heap[Parent(maxIndex)]) < 0)
+                    {
+                        SwapNodes(maxIndex, Parent(maxIndex));
+                    }
+
+                    PushDownMax(maxIndex);
+                }
+            }
+            else if (Comparer.Compare(Heap[maxIndex], Heap[index]) > 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add the missing 'else' clause.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5421646)